### PR TITLE
restartmanager: do not apply restart policy on created containers

### PR DESCRIPTION
**Pull request by runcom**
Signed-off-by: Antonio Murdaca <runcom@redhat.com>

**- What I did**

The doc at https://docs.docker.com/config/containers/start-containers-automatically/#use-a-restart-policy says that:

```
A restart policy only takes effect after a container starts successfully. In this case, starting successfully means that the container is up for at least 10 seconds and Docker has started monitoring it.
```

but that doesn't actually reflect reality:

```
$ docker create --ti --restart=always busybox

$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
07dc8a64ea87        busybox             "sh"                3 seconds ago       Created                                 hungry_volhard

$ systemctl restart docker

$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
07dc8a64ea87        busybox             "sh"                21 seconds ago      Up 1 second                             hungry_volhard
```

As stated in  the doc, the container should not be running as it was never _started_. Bringing this up for discussion though, haven't added any test yet.

@thaJeztah @cpuguy83 @vdemeester @tonistiigi PTAL

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by cpuguy83** [comment]
This doesn't quite seem like the right place to put this fix.
Was thinking https://github.com/moby/moby/blob/master/daemon/daemon.go#L323 where it can check if it's actually been started before.

**Comment by runcom** [comment]
> This doesn't quite seem like the right place to put this fix.
> Was thinking https://github.com/moby/moby/blob/master/daemon/daemon.go#L323 where it can check if it's actually been started before.

how would you do that check though?

**Comment by cpuguy83** [comment]
The container object has a `HasBeenStartedBefore` flag that should be sufficient.

**Comment by runcom** [comment]
updated and added a test

**Comment by runcom** [comment]
@cpuguy83 PTAL

File: integration/container/restart_test.go [hunk 1]
```diff
@@ -72,8 +80,10 @@ func TestDaemonRestartKillContainers(t *testing.T) {
 					assert.NilError(t, err)
 					defer client.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{Force: true})
 
-					err = client.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
-					assert.NilError(t, err)
+					if c.xStart {
+						err = client.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
+						assert.NilError(t, err)
+					}
```
Comment (anusha-ragunathan) [inline integration/container/restart_test.go]: else verify that the container doesnt start?

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM :tiger:

File: integration/container/restart_test.go (hunk 1 above)
Comment (anusha-ragunathan) [reply integration/container/restart_test.go]: My bad. I was looking for container doesnt start check after daemon restart, which already exists.

**Comment by anusha-ragunathan** [comment]
LGTM

**Comment by jaysoffian** [comment]
FWIW, this was a backwards incompatible change, despite the documentation.  IMO, the documentation should've been updated to reflect the behavior, not the other way around. I had been relying on the prior behavior, which is quite useful. In my use case, I create the container while building an AWS AMI, but don't wish to actually start it at that time. On first boot of the AMI, I then expect the container to start, which before this change it did. After this change, I need a first-boot script to start the container. Boo. :-(

**Pull request merged**